runtime/jvmti/jvmtiClass.c

```diff
@@ -1233,10 +1233,12 @@ redefineClassesCommon(jvmtiEnv* env,
 		/* Eliminate dark matter so that none will be encountered in prepareToFixMemberNames(). */
 		UDATA savedAllowUserHeapWalkFlag = vm->requiredDebugAttributes & J9VM_DEBUG_ATTRIBUTE_ALLOW_USER_HEAP_WALK;
 		vm->requiredDebugAttributes |= J9VM_DEBUG_ATTRIBUTE_ALLOW_USER_HEAP_WALK;
-		/* J9MMCONSTANT_EXPLICIT_GC_RASDUMP_COMPACT allows the GC to run while the current thread is holding
-		 * exclusive VM access.
-		 */
-		vm->memoryManagerFunctions->j9gc_modron_global_collect_with_overrides(currentThread, J9MMCONSTANT_EXPLICIT_GC_RASDUMP_COMPACT);
+		vm->alreadyHaveExclusive = TRUE;
+
+		vm->memoryManagerFunctions->j9gc_modron_global_collect_with_overrides(currentThread, J9MMCONSTANT_EXPLICIT_GC_EXCLUSIVE_VMACCESS_ALREADY_ACQUIRED);
+
+		vm->alreadyHaveExclusive = FALSE;
```

Comment:
Confirmatory question: Multiple threads can't call this code concurrently since calls to the parent function are protected by `jvmtiData->redefineMutex`?

Reply:
Not an expert in HCR, but that was my assumption all along. I don't have anything to add to your observation that seems like  the mutex should ensure it.